Introduce InstructLab Profiles Managed via `ilab profile...` to Run Key Commands at Different Fidelity Levels (#52)
instructlab/dev-docs: docs/ilab-profile.md

# Introduce Functionality to Utilize the InstructLab backend: `ilab` Profiles, and Command Adaptations

> This document describes adding different data generation, mixing, and training configurations to the `ilab` CLI to enable higher-fidelity training using the backend code. By higher-fidelity, we mean models that perform better, were trained on better hardware, off of larger data sets, and used more intensive training techniques.

Reviewer: ```suggestion
InstructLab is a workflow for a model alignment technique for adding contributions directly to Large language Models (LLMs) using multi-phased alignment training. The Instruct Lab CLI (`ilab`) currently hard codes the different capabilities (e.g. data generation, training etc.) of the workflow.

This document describes extending the CLI to make the different capabilities of the workflow to be configurable. This would enable users to configure the CLI based on:
   - Hardware specification of the user. For example, numbers of GPUs, memory etc.
   - Using capabilities which are distributed for the workflow.
```

> Users will be able to set a pre-defined profile type in their config.yaml that would enable sane defaults for higher fidelity training, generation, and evaluation. These defaults will funnel into existing and new flags for the CLI.

Reviewer: Should this be in the what you are trying to do section? This is implementation and more "how" want to do it.

> ## `ilab` Profiles

Reviewer: ```suggestion
## Proposed Design
```

Reviewer: Now introduce the profiles concept here, giving an overview like from above.

### Building off of the InstructLab Structure Redesign

Commands like: `ilab model train` will still be a single command offered to users. However, new "profiles" offered by the CLI will enable users to toggle what mode they are running `ilab model train` in.

Eventually, we will want to enable `ilab profile init` and `ilab profile set` commands that allow users to initialize a custom profile based on one of the pre-defined ones. This is not targeted for an upcoming release, though.

### Immediate Goals and Core Principles

For the near future, there will be upper level profiles that can be initialized via `ilab config init`.

Profiles should at first be static and not exposed to users. Internally, a profile would look something like:

```yaml

> profile:

Reviewer: Profile should have a `name` attribute to identify it

#### All new options will be flags with variable levels of expertise

Options that are introduced in new commands should exist in the profile AND as command line arguments via click. **Click has a way to hide arguments using an `active` option one can implement in a custom class**. So some of the new more "expert" level flags will either be hidden unless a specific profile is enabled.

#### Is this just for training?

No, adding user profiles for generation and evaluation would be ideal. The forthcoming backend libraries are taking an approach where it will be easiest to have pre-baked "config" objects that will feed into them. We need to be careful about how we introduce new concepts like this. Adding lower level intensive configurations to ilab is confusing to community users.

**PLEASE NOTE:**

**For immediate releases I would introduce the idea of a profile, a way to set these via `ilab config init` and hardcoded profiles that plug into key `ilab` commands namely training, generation, and evaluation.**

### Reasoning

> The profile settings will be used as arguments for most if not all libraries being introduced to the `ilab` backend.

Reviewer: ```suggestion
The profile settings will be used as arguments for most if not all capabilities that are part of the workflow.
```

  max_ctx_size: 4096
  model_path: models/merlinite-7b-lab-Q4_K_M.gguf
``` 
Notice I am proposing that if a profile is specified, the generation defaults move inside of the profile. This will be discussed more below.

Also, there should most likely be a default **cpu** profile always set if the user does not choose a different one.

## Proposal for Default Profiles

1. **CPU Only, Laptop Profile**
    - low instruction number for generation
    - gpus: -1 for training, eval, generation
    - no deepspeed config or accelerator config for training
    - low epoch number for training
    - single phase training, no eval support given that there are no GPUs.
2. **Single GPU, Generate, Train, Eval profile -- CUDA**
    - mid level instruction number for generation, higher than the default 10 (100-500)
    - eventual deepspeed support with LoRA and QLoRA in training
    - 10+ epochs for training
    - eval on checkpoints if hardware permits (depends on vRam)
3. **Multi GPU, Generate, Train, Eval profile -- CUDA**
    - high instruction count, 500 or so for generation
    - deepspeed support without quantization or LoRA
    - evaluation on checkpoints after a phase of training

> **Profile 3 would most likely have sub-profiles for different GPU support.**

Reviewer: Where is the Mac Metal profile covered?
